### DISCUSSION TITLE
How can i enable modules at the minimal ports?

### DISCUSSION THREAD

**Discussion by QiaoTuCodes** [micropython/micropython#9780] [Hardware & Peripherals] [answered]
a question about How to enable modules in the minimal ports? i mean how can i enable rtc and i2c or other modules to conformity in firmware ? like very small firmware customized. now testing minimal ports is good. but not anything modules in firmware no good.

**Comment by robert-hh** [comment]
Any module that uses hardware requires specific drivers. Some functions can be enabled by software. If for instance your minimal port has support for machine.pin, you can enable softspi by a config switch. If the pin module supports open drain mode, you can as well enable softI2C. RTC usually requires a specific driver. But some timekeeping may be included in the time module.

**Reply by QiaoTuCodes** [reply]
In STM32 Ports ,will cofigure the mpconfigboard.h and pin.csv files. but the 2files not in minimal port folders. so i mean , can you get me a demo or sample files teach me how to do that. thanks you. i try activate rtc and useing i2c driver ssd1306 oled to test this ports.

**Reply by jimmo** [reply]
See the STM32 port for a demo (or perhaps samd, or mimxrt, which are similar but slightly less complicated)

The minimal port just exists to show the basics of how to create a port.

**Reply by QiaoTuCodes** [reply]
> Any module that uses hardware requires specific drivers. Some functions can be enabled by software. If for instance your minimal port has support for machine.pin, you can enable softspi by a config switch. If the pin module supports open drain mode, you can as well enable softI2C. RTC usually requires a specific driver. But some timekeeping may be included in the time module.

Thank you for reply me. but i get it answer correct from Mr.Jimmo.

**Comment by jimmo** [comment, marked as answer]
Also in https://github.com/orgs/micropython/discussions/9768 you said you were using an STM32 part. Why not just use the STM32 port?

**Reply by QiaoTuCodes** [reply]
ah~ sorry . just you suggest me say here: https://github.com/micropython/micropython/issues/8786 . so i mean the big firmware must cut. so try change minimal port to useing.

**Reply by jimmo** [reply]
Yes, you should do what I suggested in #8786 ! Remove features from the stm32 port rather than trying to re-implement them on the minimal port.

Like I said, start by using the minimum feature level, then go through and remove the other peripherals and features that you don't need.

**Reply by QiaoTuCodes** [reply]
ah! i get it. so thank you very much!
